### PR TITLE
[JIT] Implement Tensor.tolist()

### PR DESCRIPTION
**Pull request by SplitInfinity**
**Summary**
This commit adds an implementation of `Tensor.tolist()` to the JIT interpreter.

**Testing**
This commit adds several unit tests that test that this function works correctly for
0D, 1D, 2D and 3D tensors of type `float`, `int` and `bool`.

```
(base) meghanl-mbp:pytorch meghanl$ python test/test_jit.py TestList.test_to_list -v
Fail to import hypothesis in common_utils, tests are not derandomized
test_to_list (jit.test_list_dict.TestList)
Unit tests for Tensor.tolist() function. ... ok

----------------------------------------------------------------------
Ran 1 test in 0.329s

OK
```

### CODE REVIEW

File: aten/src/ATen/core/interned_strings.h
Comment (SplitInfinity) [inline aten/src/ATen/core/interned_strings.h]: Not sure how that got in here, will remove.

File: torch/csrc/jit/ir.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/ir.cpp]: I don't like this either, but I have to do this to avoid capturing the output type in the operator lambda in the interpreter.

File: torch/csrc/jit/operator.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/operator.cpp]: Should this be `aten::ToList`? Does the capitalization need to be changed?

File: torch/csrc/jit/passes/python_print.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/passes/python_print.cpp]: To be honest, I guessed here, but it looks like the automatic import/export tests don't run, so this isn't being tested at the moment (cc: @eellison)

**Review by SplitInfinity — COMMENTED** [review]
Just some comments/things I noticed.

File: torch/csrc/jit/passes/python_print.cpp
Comment (eellison) [reply torch/csrc/jit/passes/python_print.cpp]: i'm trying to land https://github.com/pytorch/pytorch/pull/33468... hopefully will land soon and you can rebase on top of it.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: I'm not 100% sure but I think this should be `data += strides[dim] * elementSize`. i'm think that's what the equivalent logic in `recursive_to_list` does: https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/torch/csrc/utils/tensor_list.cpp#L26

Also: this isn't necessary, but it might make this code a little bit easier to reason about if it had a closer correspondence to `recursive_to_list`. 

`tensorToListRecursive` could return an `IValue` instead of `c10::impl::GenericList`. When `num_tensor_dim` == `cur_dim`, return the Scalar IValue, otherwise return the GenericList.  I think this would also make it so that you no longer need to specially handle 0-dim tensors in the body of the operator.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Could you add error handling for when the tensor isn't the expected type? E.g. user said List[float] but it's a non-float tensor, or it's a tensor of the wrong dims.

File: torch/csrc/jit/script/ir_emitter.cpp
Comment (eellison) [inline torch/csrc/jit/script/ir_emitter.cpp]: Can you add error handling for when no `type_hint` is passed in, ideally with a helpful error message ? Most users will be coming from python eager where they didn't have a type hint, so pretty much everyone who uses `tolist` will see this error message.

File: torch/csrc/jit/passes/alias_analysis.cpp
Comment (eellison) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Since toList doesn't mutate its input and produces a fresh alias, we can put toList with the other `analyzeCreator` cases. analyzeConservative assumes that all inputs and outputs are written to.

File: test/jit/test_list_dict.py
Comment (eellison) [inline test/jit/test_list_dict.py]: Other tests to add would be no type hint, type hint doesn't match the runtime tensor type, and type hint isn't an allowed type (non-List, non-scalar).

**Review by eellison — COMMENTED** [review]
Looks great!! Adding this op cuts across a lot of different parts of the code base and different concepts so it's a tricky one to add, but this is almost ready to land.

I think there are a couple edge-cases we need to add error handling for, and we should probably wait for my PR fixing jitter tests to land before landing this.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: It's for referring to parameter in Doxygen in running text:

http://www.doxygen.nl/manual/commands.html#cmdp

File: torch/csrc/jit/passes/python_print.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Looks like the land job failed :'(

File: torch/csrc/jit/operator.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/operator.cpp]: Done.

File: torch/csrc/jit/passes/python_print.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Fixed.

File: torch/csrc/jit/ir.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/ir.cpp]: Fixed.

File: torch/csrc/jit/passes/python_print.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/passes/python_print.cpp]: Fixed.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Fixed.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Fixed. Also, I don't think I used this anywhere so maybe I can just remove it altogether...

File: test/jit/test_list_dict.py
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Done.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done and done.

I didn't realize that transposition of a tensor adjusts the strides/sizes rather than actually transposing the data. So yes, it should be `data += strides[dim] * elementSize` even if it's the last dimension. The unit tests I added for noncontiguous tensors exposed this issue. Thanks!

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done.

File: torch/csrc/jit/script/ir_emitter.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/script/ir_emitter.cpp]: Done, although I'm not sure how helpful the message is? I wasn't sure how to get the expected type annotation at this point.

File: torch/csrc/jit/passes/alias_analysis.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Done.

File: test/jit/test_list_dict.py
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Done. I tried to add one for a tensor with an unsupported `dtype`, but I couldn't even create such a tensor lol...

File: test/jit/test_list_dict.py
Comment (eellison) [inline test/jit/test_list_dict.py]: i think you forgot to compile this one

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: maybe add test for this case

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: same for this case

**Review by eellison — APPROVED** [review]
Looks great, ship it !!!!

**Comment by eellison** [comment]
next time if you don't mind re-request review, i (& other people on the team) usually do our reviews off that.

File: test/jit/test_list_dict.py
Comment (SplitInfinity) [reply test/jit/test_list_dict.py]: Meant to delete this because I cannot create a `Tensor` of a type that `tolist` does not support

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: Done.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: Is there a reason for deconstructing the tensor here instead of just passing it along?

File: torch/csrc/jit/register_prim_ops.cpp
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: I'm not sure if it will work here but using `TensorIterator` will handle things like strides for you, instead of manually looping over the tensor's data

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: I basically copied the Python implementation and that's what that does.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/register_prim_ops.cpp]: I'm new here and I basically copied the Python implementation. If you feel strongly about this I can do it in a follow up PR.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: hey sorry, one last thing, we need to move it to CPU if it's not already 
https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/torch/csrc/utils/tensor_list.cpp#L33

**Review by eellison — COMMENTED** [review]
one last thing we forgot. i think the close equivalence to the python implementation is a good thing because it makes it easier to review / maintain. i did that with torch.tensor (equal and opposite op) and we haven't had any bugs from it.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: you can do this in a follow up as well

**Pull request closed without merging**
